File: SmartReceipts/Persistence/Migrations/DatabaseMigration.m
//
//  DatabaseMigration.m
//  SmartReceipts
//
//  Created by Jaanus Siim on 26/04/15.
//  Copyright (c) 2015 Will Baumann. All rights reserved.
//

#import "DatabaseMigration.h"
#import "Constants.h"
#import "DatabaseCreateAtVersion11.h"
#import "DatabaseUpgradeToVersion12.h"
#import "Database.h"
#import "DatabaseUpgradeToVersion13.h"
#import "Database+Functions.h"
#import "SmartReceipts-Swift.h"

@implementation DatabaseMigration

+ (NSArray *)allMigrations {
    return @[
            [DatabaseCreateAtVersion11 new],
            [DatabaseUpgradeToVersion12 new],
            [DatabaseUpgradeToVersion13 new],
            [DatabaseUpgradeToVersion14 new],
            [DatabaseUpgradeToVersion15 new],
            [DatabaseUpgradeToVersion16 new],
            [DatabaseUpgradeToVersion17 new],
            [DatabaseUpgradeToVersion18 new],
            [DatabaseUpgradeToVersion19 new],
    ];
}

- (NSUInteger)version {
    ABSTRACT_METHOD
    return 0;
}

- (BOOL)migrate:(Database *)database {
    ABSTRACT_METHOD
    return NO;
}

+ (BOOL)migrateDatabase:(Database *)database {
    NSArray *migrations = [self allMigrations];
    return [self runMigrations:migrations onDatabase:database];
}

+ (BOOL)runMigrations:(NSArray *)migrations onDatabase:(Database *)database {
    TICK;
    NSUInteger currentVersion = [database databaseVersion];
    LOGGER_INFO(@"Current version: %tu", currentVersion);

    for (DatabaseMigration *migration in migrations) {
        if (currentVersion >= migration.version) {
            LOGGER_INFO(@"DB at version %tu, will skip migration to %tu", currentVersion, migration.version);
            continue;
        }

        LOGGER_INFO(@"Migrate to version %tu", migration.version);
        if (![migration migrate:database]) {
            LOGGER_ERROR(@"Failed on migration %tu", migration.version);
            [[AnalyticsManager sharedManager] recordWithEvent:[Event failedDatabaseMigration:migration.version]];
            return NO;
        }

        currentVersion = migration.version;
        [database setDatabaseVersion:currentVersion];
    }
    
    LOGGER_DEBUG(@"Migration time %@", TOCK);
    
    return YES;
}

@end

File: SmartReceipts/Persistence/Migrations/DatabaseMigration.swift
//
//  DatabaseMigration.swift
//  SmartReceipts
//
//  Created by Bogdan Evsenev on 30/03/2019.
//  Copyright © 2019 Will Baumann. All rights reserved.
//

import Foundation
import FMDB

let DB_VERSION_UUID = 19

protocol DatabaseMigration {
    var version: Int { get }
    func migrate(_ database: Database) -> Bool
}

class DatabaseMigrator: NSObject {
    lazy var feedbackComposer: FeedbackComposer = FeedbackComposer()
    
    private static var allMigrations: [DatabaseMigration] = {
        return [
            DatabaseCreateAtVersion11(),
            DatabaseUpgradeToVersion12(),
            DatabaseUpgradeToVersion13(),
            DatabaseUpgradeToVersion14(),
            DatabaseUpgradeToVersion15(),
            DatabaseUpgradeToVersion16(),
            DatabaseUpgradeToVersion17(),
            DatabaseUpgradeToVersion18(),
            DatabaseUpgradeToVersion19()
        ]
    }()
    
    @objc func migrate(database: Database) -> Bool {
        let dbPath = database.pathToDatabase
        let copyName = "migration_failed.db"
        let copyPath = dbPath.asNSString.deletingLastPathComponent.asNSString.appendingPathComponent(copyName)
        _ = FileManager.forceCopy(from: dbPath, to: copyPath)
        let migrationResult = run(migrations: DatabaseMigrator.allMigrations, database: database)
        
        if !migrationResult {
            database.close()
            try? FileManager().removeItem(atPath: dbPath)
            _ = FileManager.forceCopy(from: copyPath, to: dbPath)
            
            processFailedMigration(databasePath: copyPath)
        } else {
            processSuccessMigration()
        }
        
        try? FileManager().removeItem(atPath: copyPath)
        return migrationResult
    }
    
    func run(migrations: [DatabaseMigration], database: Database) -> Bool {
        let tick = TickTock.tick()
        
        var currentVersion = database.databaseVersion()
        Logger.debug("Current version: \(currentVersion)")
        
        for migration in migrations {
            if currentVersion >= migration.version {
                Logger.error("DB at version \(currentVersion), will skip migration to \(migration.version)")
                continue
            }
            
            Logger.info("Migrate to version \(migration.version)")
            if !migration.migrate(database) {
                Logger.error("Failed on migration \(migration.version)")
                return false
            }
            
            currentVersion = migration.version
            database.setDatabaseVersion(currentVersion)
        }
        
        Logger.debug("Migration time \(tick.tock())")
        
        return true
    }
    
}

extension DatabaseMigrator {
    @objc static var UUIDVersion: Int = 19
}
